### PR TITLE
Solve bugs with skips and dynamic outs in stepdelegatingexecutor

### PR DESCRIPTION
**Pull request by johannkm**
This resolves 2 bugs in the step delegating executor:

- Calling is_complete after handle_event without a call to _update on ActiveExecution could return True in the middle of a job. Update the order of method calls to match the other executors
- Skip events are created and handled by the ActiveExecution. Calling handle_event with an already processed skip event results in a `Attempted to mark step {} as complete that was not known to be in flight` error. Filter out skip events from the event log before passing it up to the ActiveExecution. This wasn't discovered because if a job finishes with the skip, there's a race condition where the StepDelegatingExecutor might leave the loop before calling handle_event on the skip.

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5634**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

File: python_modules/dagster/dagster/core/execution/plan/active.py [hunk 1]
```diff
@@ -402,7 +402,9 @@ def handle_event(self, dagster_event: DagsterEvent) -> None:
         elif dagster_event.is_step_success:
             self.mark_success(step_key)
         elif dagster_event.is_step_skipped:
-            self.mark_skipped(step_key)
+            raise DagsterInvariantViolationError(
```
Comment (johannkm) [inline python_modules/dagster/dagster/core/execution/plan/active.py:405]: This is aggresive, but my current understanding is that no skip events originate outside of the ActiveExecution, and so this should never be called. If that's not the case, then the StepDelegatingExecutor would have a hard to diagnose bug.

File: python_modules/dagster/dagster/core/executor/step_delegating/step_delegating_executor.py [hunk 2]
```diff
@@ -127,6 +127,10 @@ def execute(self, plan_context: PlanOrchestrationContext, execution_plan: Execut
                     running_steps[step.key] = step
 
             last_check_step_health_time = pendulum.now("UTC")
+
+            # Order of events is important here. During an interation, we call handle_event, then get_steps_to_execute,
+            # then is_complete. get_steps_to_execute updates the state of ActiveExecution, and without it
+            # is_complete can return true when we're just between steps.
```
Comment (johannkm) [inline python_modules/dagster/dagster/core/executor/step_delegating/step_delegating_executor.py:133]: Could also consider modifying the ActiveExecution...

File: python_modules/dagster/dagster/core/execution/plan/active.py (hunk 1 above)
Comment (gibsondan) [reply python_modules/dagster/dagster/core/execution/plan/active.py:405]: maybe worth a comment here explaining for people who don't have the context of the diff summary

**Review by gibsondan — APPROVED** [review]
awesome - where did we land on the flaky IOManager thing, that was the last remaining piece of the known executor issues I think?

**Pull request merged**